### PR TITLE
Better lifetime error message

### PR DESCRIPTION
**Pull request by mark-i-m**
I propose the following error message as more user-friendly

r? @nikomatsakis

### CODE REVIEW

File: src/librustc_mir/borrow_check/nll/region_infer/error_reporting/mod.rs [hunk 1]
```diff
@@ -517,7 +517,7 @@ impl<'tcx> RegionInferenceContext<'tcx> {
     ) {
         let mut diag = infcx.tcx.sess.struct_span_err(
             span,
-            "unsatisfied lifetime constraints", // FIXME
+            "unable to prove references are valid"
```
Comment (Centril) [inline src/librustc_mir/borrow_check/nll/region_infer/error_reporting/mod.rs]:
```suggestion
            "unable to prove that references are valid"
```

**Comment by Centril** [comment]
Hmm... Is this necessarily about proving that *references* are valid, or whether the lifetimes are the correct ones? The new error message seems less accurate in some cases that what we have right now.

**Comment by mark-i-m** [comment]
An unsatisfied lifetime constraint usually (always?) indicates that there may be a reference somewhere that may cause memory unsafety (e.g. by living too long). Or am I missing something?

**Comment by mark-i-m** [comment]
Hmm... I can see your point, but I'm not sure how to fix it. Should we make the message more vague (e.g. "unable to prove memory safety")?

**Comment by Centril** [comment]
@mark-i-m My point is that the current diagnostic talks about *lifetimes* whereas the proposed one talks about *references*; talking about lifetimes seems more direct here; could we reword it to retain lifetimes somehow but still make it more friendly?

For example:

> unable to use a lifetime which fits the expected one

This one is probably not great, but do you see what I mean?

**Comment by mark-i-m** [comment]
How about some of the following?

> lifetime may not live long enough

> lifetime may live too long

> inferred lifetime constraint must be explicit

> implementation does not satisfy lifetime restrictions from function header

> bro, you're lifetimes are all wonky

**Comment by Centril** [comment]
Of the top of my head I like:

> implementation does not satisfy lifetime restrictions from function header

(assuming there is a function header...? ^.^)

but... I have sucked up too much air; I'll defer to @estebank's judgement on diagnostics :)

**Comment by mark-i-m** [comment]
> My ideal solution would be to rework the message depending on which of the possible situations we're facing. For example, I'd like it to be something along the lines of "borrow being returned doesn't live long enough".

So we do that currently, but this is the fallback message if we can do one of the more specialized messages.

**Comment by nikomatsakis** [comment]
In general, it would be helpful to me if you would run `./x.py test --bless` so we can judge the impact of the message in context.

**Comment by mark-i-m** [comment]
@nikomatsakis Pushed.

**Comment by estebank** [comment]
We need to keep iterating over these error messages, but I'm happy with this PR as is as I find the new wording less jargon-ny. @nikomatsakis what do you think?

**Review by estebank — APPROVED** [review]

**Comment by mark-i-m** [comment]
I don't understand the CI failure. Everything is passing locally. Don't I just need to rebase or something?

**Comment by estebank** [comment]
You need to update https://github.com/rust-lang/rust/blame/master/src/test/ui/nll/user-annotations/issue-54124.rs (you did update https://github.com/rust-lang/rust/blame/master/src/test/ui/nll/user-annotations/issue-54124.stderr) and https://github.com/rust-lang/rust/blame/master/src/test/ui/nll/issue-55394.rs

**Comment by mark-i-m** [comment]
Or maybe I just missed it or did a weird rebase or something

**Comment by nikomatsakis** [comment]
I'm torn. I agree the new message has less jargon. On the other hand, I'm not really sure what a lifetime "living long enough" should mean, and I'm interested in seeing what we can do to move away from this "lifetime" and "live long enough" sort of terminology, which I think is generally sort of confusing.

Still, I don't really have any superior suggestions. =)

**Comment by Centril** [comment]
Perhaps we can talk about "Object `x` is valid for the lifetime `'a`, the lifetime `'b` was expected; but 'a is shorter than `'b`."; not sure how to turn it into prose... Perhaps one of:
+ "error: lifetime is shorter than required"
+ "error: object is not valid for the expected lifetime"

**Comment by mark-i-m** [comment]
> I'm not really sure what a lifetime "living long enough" should mean

In my mind, this means their is a reference somewhere that we can't prove is valid everywhere it is used, which is what I put in the original change in this PR. Something about the intuition of "X might not be valid everywhere it is used" seems like the intuition we want to convey IMHO.

**Comment by mark-i-m** [comment]
I'm not sure what the status of this is. Is it waiting for me or someone else?

**Comment by Centril** [comment]
I'll say that it is waiting on @nikomatsakis but maybe they want to throw it back at you... =P

**Comment by estebank** [comment]
If there are further changes needed, they can be addressed in a follow up PR.

**Comment by Centril** [comment]
Failed in https://github.com/rust-lang/rust/pull/57688#issuecomment-455039689 when combined with other things (not sure what...). I think you need to `--bless --compare-mode=nll`.

@bors r-

**Comment by mark-i-m** [comment]
I have to go now, and I haven't had a chance to run the full suite with --bless --compare-mode=nll, but the ui tests are updated and the rest are running now.

**Comment by Centril** [comment]
Ugh... seems we got a segfault (I would be surprised if this PR was the cause, but still worrying...)?

```java
[01:46:20] test vec::test_swap_remove_empty ... ok
[01:46:21] died due to signal 11
[01:46:21] test failed, to rerun pass '--test collectionstests'
```

@bors retry
